astropy/io/fits/connect.py

```diff
@@ -265,21 +265,31 @@ def read_table_fits(
         # string, empty strings.
         # Since Multi-element columns with dtypes such as '2f8' have a subdtype,
         # we should look up the type of column on that.
+        # Also propagate TNULL (for ints) or the FITS default null value for
+        # floats and strings to the column's fill_value to ensure round trips
+        # preserve null values.
         masked = mask = False
         coltype = col.dtype.subdtype[0].type if col.dtype.subdtype else col.dtype.type
         if col.null is not None:
             mask = data[col.name] == col.null
             # Return a MaskedColumn even if no elements are masked so
             # we roundtrip better.
             masked = True
+            fill_value = col.null
         elif mask_invalid and issubclass(coltype, np.inexact):
             mask = np.isnan(data[col.name])
+            fill_value = np.NaN
```

Comment:
Are this and the line below covered in tests?

Reply:
I've extended `test_null_propagation_in_table_read` to ensure that float/complex/string columns have the expected `fill_value` (see 59d7256863a5589589edb7bc2d9824bf65f951cf)